internal/transport/transport_test.go

```diff
@@ -3008,6 +3008,50 @@ func (s) TestClientCloseReturnsEarlyWhenGoAwayWriteHangs(t *testing.T) {
 	ct.Close(errors.New("manually closed by client"))
 }
 
+// deadlineTestConn is a net.Conn wrapper used to assert that deadlines are set
+// during http2Client.Close().
+type deadlineTestConn struct {
+	net.Conn
+	observedReadDeadline  atomic.Bool
+	observedWriteDeadline atomic.Bool
```

Comment:
I believe we need these atomics because `t.Close` is being called after unblocking the reader of the `readerDone` channel here:
https://github.com/grpc/grpc-go/blob/3c743c9d3f630d12ec805055c304cc30fa6ea89d/internal/transport/http2_client.go#L1637-L1642

Using an atomic will satisfy the race detector, but will make the test flake if the deadline assertion code executed before the deadline is set.

@eshitachandwani is there a reason to close `readerDone` before closing the transport? I tried moving `close(t.readerDone)` below and some tests started failing. I'm not sure if that's a real deadlock of an issue in the tests.

Reply:
For now, to avoid flakiness, we can use channels instead of atomics. Instead of setting atomics, we can close the channels. In the main test function we can use a select statement to wait on the `observedReadDeadline` and `ctx.Done()` events. if `ctx.Done()` happens, we should fail the test, e.g:
https://github.com/grpc/grpc-go/blob/3c743c9d3f630d12ec805055c304cc30fa6ea89d/internal/transport/transport_test.go#L817-L821

Similarly we can add a select statement for `observedWriteDeadline`.

Reply:
Oh, I see. We're calling `t.Close` twice. It is guaranteed that the deadline is set after `t.Close` returns, the reader goroutine may call it again, but it will be a no-op. The use of atomics here is fine.